### PR TITLE
LEXI-88 Create ClayBadge

### PR DESCRIPTION
**Pull request by matuzalemsteles**

### CODE REVIEW

**Comment by brunobasto** [comment]
@matuzalemsteles can you put the ticket number on the commit message?

File: packages/clay-badge/src/ClayBadge.soy [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+	{let $classes kind="text"}
+		badge badge-pill
+		{sp}badge-{$style ?: 'default'}
+
+		{if $size}
+			{sp}badge-{$size}
+		{/if}
+	{/let}
+
+	{let $attributes kind="attributes"}
```
Comment (brunobasto) [inline packages/clay-badge/src/ClayBadge.soy]: This seems to only be necessary inside the `else` statement, right?

File: packages/clay-badge/src/ClayBadge.soy [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+/**
+ * Renders icon content
+ */
+{template .content}
+	{@param? closeable: bool}
+	{@param? label: html|string}
+	{@param? spritemap: string}
+
+	{$label ? $label : ''}
```
Comment (brunobasto) [inline packages/clay-badge/src/ClayBadge.soy]: I don't see where this `label` param is being used. Do we need it?

File: packages/clay-badge/src/ClayBadge.soy [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+	{if $closeable and $spritemap}
+		<button aria-label="Close" class="close" type="button">
```
Comment (brunobasto) [inline packages/clay-badge/src/ClayBadge.soy]: How are we localizing hard-coded strings?

**Review by brunobasto — CHANGES_REQUESTED** [review]

**Comment by matuzalemsteles** [comment]
> can you put the ticket number on the commit message?

@brunobasto yes, do you think we should put it right in the `span` tag?

File: packages/clay-badge/src/ClayBadge.soy (hunk 1 above)
Comment (matuzalemsteles) [reply packages/clay-badge/src/ClayBadge.soy]: Yes, do you think we should put it right in the`span` tag?

File: packages/clay-badge/src/ClayBadge.soy (hunk 2 above)
Comment (matuzalemsteles) [reply packages/clay-badge/src/ClayBadge.soy]: Yes, when we do not pass `href` and we want to render only the `label` with icon or without.

File: packages/clay-badge/src/ClayBadge.soy (hunk 3 above)
Comment (matuzalemsteles) [reply packages/clay-badge/src/ClayBadge.soy]: We can locate by the metal selector, for example:
```javascript
events: {
	click: {
		selector: '.close',
		fn: handleClick
	}
}
```

**Comment by matuzalemsteles** [comment]
Hey @brunobasto,

> can you put the ticket number on the commit message?

Yes, consider it done.

**Comment by carloslancha** [comment]
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

**Comment by carloslancha** [comment]
Ey @matuzalemsteles I'm reviewing [Lexicon's Badges documentation](https://docs.google.com/document/d/1ZU2UoH-MhGDH2fOsW-WyjLF3JD5iS2naMPI4KHDOd1U/edit) (attached in the ticket) and I've noticed that
- A badge cannot be a link
- A badge cannot be closeable

Don't know why I didn't see that reviewing your last pull, maybe I was looking in the wrong place. Sorry about that 😓 

If you have any doubt do not hesitate to ask me or Lexicon team

File: packages/clay-badge/src/ClayBadge.soy (hunk 2 above)
Comment (brunobasto) [reply packages/clay-badge/src/ClayBadge.soy]: Ah .Got it

File: packages/clay-badge/src/ClayBadge.soy (hunk 3 above)
Comment (brunobasto) [reply packages/clay-badge/src/ClayBadge.soy]: I meant `aria-label="Close"` will not be translated. Are there other components where we have hard-coded strings? If so, what's the pattern we are following?

Comment (carloslancha) [reply packages/clay-badge/src/ClayBadge.soy]: Talking with @jbalsas we think we should go with it hardcoded for now while we find the best solution. It's an interesting issue...

Comment (brunobasto) [reply packages/clay-badge/src/ClayBadge.soy]: Yeah :)

In portal, we have a mechanism in place that will translate keys used with
```soy
{msg desc=""}lang-key{/msg}
```

Not sure we could use it here, though

Comment (carloslancha) [reply packages/clay-badge/src/ClayBadge.soy]: Yep, the point is that if we decide to go in that direction we should provide some "easy" mechanism to allow users to pass the keys to the component, or a series of .json with them that the build process can use... I don't know, just thinking loud... :P

**Comment by carloslancha** [comment]
@matuzalemsteles I've talk with Victor Valle about the abbreviation stuff. This abbrs are not international, 100K could have no meaning for some cultures or 1M have different ones, so we decided to keep this as simple as possible and forget about the abbrs, leaving that decision to the user of the component. So Label should be a string.

Anyway, just as a hint for future developments, we need to have in mind both client and server renders. If we massage certain data in the JS (like here with the abbr) we have to replicate that behaviour in the soy in order to make both renders consistent.

Thx!

File: packages/clay-badge/src/ClayBadge.soy [hunk 4]
```diff
@@ -10,7 +10,7 @@
 
 	{let $classes kind="text"}
 		badge badge-pill
-		{sp}badge-{$style ?: 'default'}
+		{sp}badge-{$style}
```
Comment (carloslancha) [inline packages/clay-badge/src/ClayBadge.soy]: If default value is `primary` as defined in the STATE this should be `{sp}badge-{$style ?: 'primary'}`

File: packages/clay-badge/src/ClayBadge.js [hunk 5]
```diff
@@ -73,10 +30,10 @@ ClayBadge.STATE = {
 	 * The label of the badge element.
 	 * @instance
 	 * @memberof ClayBadge
-	 * @type {number}
+	 * @type {number|string}
 	 * default undefined
 	 */
-	label: Config.number().setter('hasNumberLabel_').required(),
+	label: Config.any().required(),
```
Comment (carloslancha) [inline packages/clay-badge/src/ClayBadge.js]: `Config.any()` accepts anything, including functions, arrays, etc. This should be just string.

**Review by carloslancha — CHANGES_REQUESTED** [review]

File: packages/clay-badge/src/ClayBadge.js [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+	 * @type {?string|undefined}
+	 * default undefined
+	 */
+	id: Config.string(),
+
+	/**
+	 * The label of the badge element.
+	 * @instance
+	 * @memberof ClayBadge
+	 * @type {string}
```
Comment (carloslancha) [inline packages/clay-badge/src/ClayBadge.js]: If required `@type` should be preceded by `!`

`@type {!string}`

File: packages/clay-badge/src/ClayBadge.js [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+	/**
+	 * Badge style color.
+	 * @instance
+	 * @memberof ClayBadge
+	 * @type {?string|undefined}
+	 * default undefined
+	 */
+	style: Config.oneOf([
+		'primary',
```
Comment (carloslancha) [inline packages/clay-badge/src/ClayBadge.js]: Order alphabetically

File: packages/clay-badge/src/ClayBadge.soy [hunk 8]
```diff
@@ -0,0 +1,20 @@
+{namespace ClayBadge}
+
+/**
+ * This renders a badge.
+ */
+{template .render}
+	{@param? id: string}
```
Comment (carloslancha) [inline packages/clay-badge/src/ClayBadge.soy]: Order params

File: packages/clay-badge/src/ClayBadge.soy [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+
+	{let $classes kind="text"}
+		badge badge-pill
+		{sp}badge-{$style ?: 'primary'}
+	{/let}
+
+	<span class="{$classes}" id="{$id}">
+		{$label}
+	</span>
+
```
Comment (carloslancha) [inline packages/clay-badge/src/ClayBadge.soy]: No empty line needed.

**Review by carloslancha — CHANGES_REQUESTED** [review]
Could you please rebase with master and run `npm run mcritic` to confirm that passes validation?

Thx!

File: packages/clay-badge/src/ClayBadge.js [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+	 */
+	label: Config.string().required(),
+
+	/**
+	 * Badge style color.
+	 * @instance
+	 * @memberof ClayBadge
+	 * @type {?string|undefined}
+	 * default undefined
+	 */
```
Comment (carloslancha) [inline packages/clay-badge/src/ClayBadge.js:44]: default is primary, not undefined

Comment (matuzalemsteles) [reply packages/clay-badge/src/ClayBadge.js:44]: You're right.

**Comment by carloslancha** [comment]
LGTM Thx!!

**Pull request merged**